Add an rpath flag to disable the default of looking in the local directory for shared libraries

finagolfin opened the pull request:
This is useful when deploying build products, as most executables are not installed in the same directory as libraries and system libraries do not need it for system paths. Use the new flag when installing SwiftPM itself and the {Manifest,Plugin}API libraries on ELF platforms.

On linux, I currently see this with Swift 5.9:
```
> readelf -d swift-5.9-RELEASE-ubuntu20.04/usr/lib/swift/pm/*I/lib*so swift-5.9-RELEASE-ubuntu20.04/usr/bin/swift-package|ag "File:|runpath"
File: swift-5.9-RELEASE-ubuntu20.04/usr/lib/swift/pm/ManifestAPI/libPackageDescription.so
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN:$ORIGIN/../../linux]
File: swift-5.9-RELEASE-ubuntu20.04/usr/lib/swift/pm/PluginAPI/libPackagePlugin.so
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN:$ORIGIN/../../linux]
File: swift-5.9-RELEASE-ubuntu20.04/usr/bin/swift-package
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN:$ORIGIN/../lib/swift/linux]
```
After applying this pull to trunk on Android, the extraneous rpath is gone:
```
> llvm-readelf -d build/Ninja-Release/toolchain-android-aarch64/usr/lib/swift/pm/*I/lib*so build/Ninja-Release/toolchain-android-aarch64/usr/bin/swift-package|ag "File:|runpath"
File: build/Ninja-Release/toolchain-android-aarch64/usr/lib/swift/pm/ManifestAPI/libPackageDescription.so
  0x000000000000001d (RUNPATH)      Library runpath: [$ORIGIN/../../android]
File: build/Ninja-Release/toolchain-android-aarch64/usr/lib/swift/pm/PluginAPI/libPackagePlugin.so
  0x000000000000001d (RUNPATH)      Library runpath: [$ORIGIN/../../android]
File: build/Ninja-Release/toolchain-android-aarch64/usr/bin/swift-package
  0x000000000000001d (RUNPATH)         Library runpath: [$ORIGIN/../lib/swift/android:/data/data/com.termux/files/usr/lib]
```
I have no idea if this is useful on Darwin too, but I went ahead and disabled adding `@loader_rpath` too. Let me know if that's worthwhile.

finagolfin commented on Sources/CoreCommands/Options.swift:
```diff
@@ -524,6 +524,10 @@ public struct LinkerOptions: ParsableArguments {
+
+    /// Disables adding $ORIGIN/@loader_path to the rpath, useful when deploying
+    @Flag(name: .customLong("disable-local-rpath"), help: "Disable adding $ORIGIN/@loader_path to the rpath by default")
+    public var shouldDisableLocalRpath: Bool = false
```
I don't care what the flag name is, feel free to suggest another.

neonichu replied:
I would call it "origin" instead of "local"

finagolfin replied:
I considered that originally but used this broader term when I disabled the Darwin `@loader_path` too. Do you want me to keep that disabled for Darwin and change this flag and variable name to `origin`? I have no preference on these two aspects of this pull.

finagolfin commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -5021,4 +5021,87 @@ final class BuildPlanTests: XCTestCase {
         }
     }
+
+    func testBasicSwiftPackageWithoutLocalRpath() throws {
```
This test is a copy and paste of `testBasicSwiftPackage()` above, with the rpath disabled and a few checks removed.

finagolfin commented:
@neonichu, please run the CI on this.

neonichu commented:
@swift-ci please test

finagolfin commented:
Linux CI passed, while Windows CI failed spuriously.

Mac CI said it didn't find a warning diagnostic from the old test I modified, so I just updated the test to not look for that.

finagolfin commented:
@neonichu, please rerun the CI.

finagolfin commented:
@bnbarham, would you run the CI on this?

bnbarham commented:
@swift-ci please test

finagolfin commented:
@neonichu, passed CI, ready for review.
